File: strings/z_function.go
package main

import (
	"fmt"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func ZFunction(s string) []int {
	n := len(s)
	z := make([]int, n)
	l, r := 0, 0

	for i := 1; i < n; i++ {
		if i <= r {
			z[i] = min(r-i+1, z[i-l])
		}
		for i+z[i] < n && s[z[i]] == s[i+z[i]] {
			z[i]++
		}
		if i+z[i]-1 > r {
			l = i
			r = i + z[i] - 1
		}
	}
	return z
}

func FindPattern(pattern, text string) int {
	combined := pattern + "$" + text
	z := ZFunction(combined)
	count := 0
	pLen := len(pattern)

	for _, val := range z {
		if val == pLen {
			count++
		}
	}
	return count
}

func main() {
	text := "abracadabra"
	pattern := "abr"
	fmt.Printf("Occurrences: %d\n", FindPattern(pattern, text))
}

File: strings/z_function.rs
use std::cmp;

pub fn z_function(s: &str) -> Vec<usize> {
    let n = s.len();
    let s_bytes = s.as_bytes();
    let mut z = vec![0; n];
    let mut l = 0;
    let mut r = 0;

    for i in 1..n {
        if i <= r {
            z[i] = cmp::min(r - i + 1, z[i - l]);
        }
        while i + z[i] < n && s_bytes[z[i]] == s_bytes[i + z[i]] {
            z[i] += 1;
        }
        if i + z[i] - 1 > r {
            l = i;
            r = i + z[i] - 1;
        }
    }
    z
}

pub fn find_pattern(pattern: &str, text: &str) -> i32 {
    let combined = format!("{}${}", pattern, text);
    let z = z_function(&combined);
    let mut count = 0;
    let p_len = pattern.len();

    for &val in z.iter() {
        if val == p_len {
            count += 1;
        }
    }
    count
}

fn main() {
    let text = "abracadabra";
    let pattern = "abr";
    println!("Occurrences: {}", find_pattern(pattern, text));
}
